src_lockfree/mpsc_queue.mli

```diff
@@ -1,69 +1,67 @@
 (** Lock-free multi-producer, single-consumer, domain-safe queue
-    without support for cancellation.
-
-    This makes a good data structure for a scheduler's run queue and
-    is currently (September 2022) used for Eio's scheduler. *)
-
-type 'a t
-(** A queue of items of type ['a]. *)
-
-exception Closed
-
-val create : unit -> 'a t
-(** [create ()] returns a new empty queue. *)
-
-val is_empty : 'a t -> bool
-(** [is_empty q] is [true] if calling [pop] would return [None].
-
-    @raise Closed if [q] is closed and empty. *)
-
-val close : 'a t -> unit
-(** [close q] marks [q] as closed, preventing any further items from
-    being pushed by the producers (i.e. with {!push}).
-
-    @raise Closed if [q] has already been closed. *)
-
-val push : 'a t -> 'a -> unit
-(** [push q v] adds the element [v] at the end of the queue [q]. This
-    can be used safely by multiple producer domains, in parallel with
-    the other operations.
-
-    @raise Closed if [q] is closed. *)
-
-(** {2 Consumer-only functions} *)
-
-exception Empty
-(** Raised when {!pop} or {!peek} is applied to an empty queue. *)
-
-val pop : 'a t -> 'a
-(** [pop q] removes and returns the first element in queue [q].
-
-    @raise Empty if [q] is empty.
-
-    @raise Closed if [q] is closed and empty. *)
-
-val pop_opt : 'a t -> 'a option
-(** [pop_opt q] removes and returns the first element in queue [q] or
-    returns [None] if the queue is empty.
-
-    @raise Closed if [q] is closed and empty. *)
-
-val peek : 'a t -> 'a
-(** [peek q] returns the first element in queue [q].
-
-    @raise Empty if [q] is empty.
-
-    @raise Closed if [q] is closed and empty. *)
-
-val peek_opt : 'a t -> 'a option
-(** [peek_opt q] returns the first element in queue [q] or
-    returns [None] if the queue is empty.
-
-    @raise Closed if [q] is closed and empty. *)
-
-val push_head : 'a t -> 'a -> unit
-(** [push_head q v] adds the element [v] at the head of the queue
-    [q]. This can only be used by the consumer (if run in parallel
-    with {!pop}, the item might be skipped).
-
-    @raise Closed if [q] is closed and empty. *)
+    without support for cancellation.*)
+
+    type 'a t
+    (** A queue of items of type ['a]. *)
+    
+    exception Closed
+    
+    val create : unit -> 'a t
+    (** [create ()] returns a new empty queue. *)
+    
+    val is_empty : 'a t -> bool
+    (** [is_empty q] is [true] if calling [pop] would return [None].
+    
+        @raise Closed if [q] is closed and empty. *)
+    
+    val close : 'a t -> unit
+    (** [close q] marks [q] as closed, preventing any further items from
+        being pushed by the producers (i.e. with {!push}).
```

Comment:
Note that with the new implementation `close` can only be safely called by the consumer.  Basically any operation whose internal implementation touches the `head` can only be safely called by the consumer.  It would be useful to e.g. group the operations in this signature file into two sections where one section contains operations that may only be called by the consumer and the other section contains operations that can also be called by the producers.

Reply:
Note that, for example, the comment at https://github.com/ocaml-multicore/saturn/blob/e3902f51a0fd8fa950c1a99502955714d76eae76/test/mpsc_queue/qcheck_mpsc_queue.ml#L3-L13 no longer holds and should be updated.